featurebyte/service/table_columns_info.py

```diff
@@ -37,6 +37,146 @@
+        target_columns_info: List[ColumnInfo]
+            Target columns info
+
+        Yields
+        ------
+        Tuple[ColumnInfo, Optional[ObjectId]]
+            Affected columns info with the original entity ID
+        """
+        original_col_to_entity_id = {
+            col_info.name: col_info.entity_id for col_info in original_columns_info
+        }
+        for col_info in target_columns_info:
+            if col_info.entity_id != original_col_to_entity_id.get(col_info.name):
+                yield col_info, original_col_to_entity_id.get(col_info.name)
+
+    async def validate_entity_dtype(
+        self, table: TableModel, target_columns_info: List[ColumnInfo]
+    ) -> None:
+        """
+        Validate entity dtype for the target columns info.
+
+        Parameters
+        ----------
+        table: TableModel
+            TableModel object
+        target_columns_info: List[ColumnInfo]
+            Target columns info
+
+        Raises
+        ------
+        DocumentUpdateError
+            When the entity dtype does not match the column dtype
+        """
+        for col_info, _ in self.iterate_affected_target_entity_columns_info(
+            original_columns_info=table.columns_info, target_columns_info=target_columns_info
+        ):
+            if col_info.entity_id:
+                # when the entity is associated with a column, validate the entity dtype
+                entity = await self.maybe_initialize_entity_dtype(col_info.entity_id)
+                if entity.dtype and col_info.dtype != entity.dtype:
+                    raise DocumentUpdateError(
+                        f"Column {col_info.name} (Table ID: {table.id}, Name: {table.name}) "
+                        f"has dtype {col_info.dtype} which does not match the entity dtype {entity.dtype}."
+                    )
+
+    async def update_entity_dtype(
+        self, table: TableModel, target_columns_info: List[ColumnInfo]
+    ) -> None:
+        for col_info, prev_entity_id in self.iterate_affected_target_entity_columns_info(
+            original_columns_info=table.columns_info, target_columns_info=target_columns_info
+        ):
+            if col_info.entity_id:
+                # when the entity does not have dtype, update the entity dtype
+                entity = await self.entity_service.get_document(col_info.entity_id)
+                if entity.dtype is None:
+                    await self.entity_service.update_document(
+                        document_id=col_info.entity_id,
+                        data=EntityServiceUpdate(dtype=col_info.dtype),
+                        return_document=False,
+                    )
+
+            if prev_entity_id and col_info.entity_id is None:
+                # when the entity is disassociated from a column, update the entity dtype if needed
+                entity = await self.entity_service.get_document(prev_entity_id)
+                if not entity.table_ids:
```

Comment:
The `table_ids` of the entity are updated first before this part, is that right?

Reply:
Yes, `table_ids` is updated before calling this. I should update the doc string here to make it clearer.